cmd/infra/powervs/create.go

```diff
@@ -1438,3 +1470,128 @@ func attachTag(gtag *globaltaggingv1.GlobalTaggingV1, infraID string, resourceId
 
 	return nil
 }
+
+// setupTransitGateway set up the transit gateway for the cluster infra
+func (infra *Infra) setupTransitGateway(ctx context.Context, options *CreateInfraOptions, gtag *globaltaggingv1.GlobalTaggingV1) error {
+	log(options.InfraID).Info("Setting up Transit Gateway ...")
+
+	tgapisv1, err := transitgatewayapisv1.NewTransitGatewayApisV1(&transitgatewayapisv1.TransitGatewayApisV1Options{
+		Authenticator: getIAMAuth(),
+		Version:       utilpointer.String(currentDate),
+	})
+
+	if err != nil {
+		return err
+	}
+	var transitGateway *transitgatewayapisv1.TransitGateway
+	if options.TransitGateway != "" {
+		log(options.InfraID).Info("Validating Transit Gateway", "name", options.TransitGateway)
+		transitGateway, err = validateTransitGatewayByName(ctx, tgapisv1, options.TransitGateway, true)
+		if err != nil {
+			return fmt.Errorf("error validating transit gateway by name %s, %w", options.TransitGateway, err)
+		}
+	} else {
+		transitGateway, err = infra.createTransitGateway(ctx, options, tgapisv1)
+		if err != nil {
+			return fmt.Errorf("error creating transit gateway: %w", err)
+		}
+	}
+
+	if transitGateway != nil {
+		infra.TransitGatewayID = *transitGateway.ID
+		infra.Stats.TransitGatewayState.Status = *transitGateway.Status
+	} else {
+		return fmt.Errorf("unable to setup transit gateway")
+	}
+
+	if err = attachTag(gtag, options.InfraID, transitGateway.Crn, fmt.Sprintf("%s-%s", infra.ID, transitGatewayNameSuffix)); err != nil {
+		log(options.InfraID).Error(err, "error attaching tags to transit gateway")
+	}
+
+	log(options.InfraID).Info("Transit Gateway Ready", "id", infra.TransitGatewayID)
+	return nil
+}
+
+// createTransitGateway creates transit gateway and connections
+func (infra *Infra) createTransitGateway(ctx context.Context, options *CreateInfraOptions, tgapisv1 *transitgatewayapisv1.TransitGatewayApisV1) (*transitgatewayapisv1.TransitGateway, error) {
+
+	transitGatewayName := fmt.Sprintf("%s-%s", options.InfraID, transitGatewayNameSuffix)
+
+	tg, err := validateTransitGatewayByName(ctx, tgapisv1, transitGatewayName, true)
+	if err != nil && err.Error() != transitGatewayNotFound(transitGatewayName).Error() {
+		return nil, fmt.Errorf("error validating existing transit gateway: %w", err)
+	}
+	if tg != nil && *tg.Status == "available" {
```

Comment:
Is there a scenario where tg is not nil and status is not in `available` state then what happens? Ideally even a create call also should fail as tg already exists, then how do we handle that scenario?

Reply:
Yes, when `validateTransitGatewayByName` returns transitGatewayNotFound error, tg would be nil, so before accessing `*tg.status`, need to validate it's not nil. If its nil, will proceed to create a new transit gatway. So we need a nil check here. We will not attempt the create if its already exists.

Reply:
what if tg is not null(transit gateway exist) and state is not `available`?